src/platform/posix/tcp_ip_channel.c

```diff
@@ -89,11 +154,53 @@ static lf_ret_t TcpIpChannel_try_connect(NetworkChannel *untyped_self) {
 
     int ret = connect(self->fd, (struct sockaddr *)&serv_addr, sizeof(serv_addr));
     if (ret < 0) {
-      return LF_COULD_NOT_CONNECT;
+      if (errno == EINPROGRESS) {
+        self->client_connect_in_progress = true;
+        LF_DEBUG(NET, "Connection in progress!");
+        return LF_TRY_AGAIN;
+      } else {
+        LF_ERR(NET, "Connect failed errno=%d", errno);
+        self->client_connect_in_progress = false;
+        TcpIpChannel_reset_socket(self);
+        return LF_TRY_AGAIN;
+      }
+    }
+  } else {
+    // Connection is in progress
+    lf_ret = check_if_socket_is_writable(self->fd);
+    if (lf_ret == LF_OK) {
+      LF_DEBUG(NET, "Socket is writable");
+      lf_ret = check_if_connect_succeeded(self->fd);
+      if (lf_ret == LF_OK) {
+        LF_DEBUG(NET, "Connection succeeded");
+        self->client_connect_in_progress = false;
+        return LF_OK;
+      } else {
+        self->client_connect_in_progress = false;
+        LF_ERR(NET, "Connection failed");
+        TcpIpChannel_reset_socket(self);
+        return LF_TRY_AGAIN;
+      }
+    } else if (lf_ret == LF_TIMEOUT) {
+      LF_ERR(NET, "Select timed out");
+      return LF_TRY_AGAIN;
+    } else {
+      self->client_connect_in_progress = false;
+      TcpIpChannel_reset_socket(self);
+      LF_ERR(NET, "Select failed errno=%d", errno);
+      return LF_TRY_AGAIN;
     }
   }
+  return LF_ERR; // Should never reach here
+}
 
-  return LF_OK;
+static lf_ret_t TcpIpChannel_try_connect(NetworkChannel *untyped_self) {
+  TcpIpChannel *self = (TcpIpChannel *)untyped_self;
+  if (self->server) {
+    return TcpIpChannel_try_connect_server(untyped_self);
```

Comment:
Since the `try_connect_server` and `try_connect_client` functions are only called within this function it could make sens to make them inline functions to allow for compiler optimizations?

Reply:
The same could also be said for all the other helper functions maybe